api/src/license_operator.rs

```diff
@@ -0,0 +1,207 @@
+        // Validate user parameters
+        // Minimal 8 seed size, USER_PAYLOAD payload size and 4 checksum size
+        let license_key_required_size: usize = 8 + self.magic.payload_size() + 4;
+        if self.key_size <= license_key_required_size {
+            info!(
+                "Cannot generate license key with less than {} key size! [key_size={}]",
+                license_key_required_size, self.key_size
+            );
+            return LicenseKey::default();
+        }
+        let license_key_hash_size = self.key_size - 4 - self.magic.payload_size();
+
+        let mut license_key = LicenseKey::default();
+        let mut serialized_license_key = Vec::new();
+
+        // Hash seed to license key
+        let mut hasher = Shake256::default();
+        hasher.update(seed);
+        let mut reader = hasher.finalize_xof();
+        let mut buffer = vec![0u8; license_key_hash_size];
+        reader.read(&mut buffer);
+
+        for byte in buffer.to_vec().iter() {
+            serialized_license_key.push(*byte);
+            license_key.key.push(*byte);
+        }
+        license_key.properties.key_size = buffer.len();
+
+        // Generate payload
+        for m in self.magic.get_magic().iter() {
+            let payload = self.serializer.hash(buffer.borrow(), m);
+            serialized_license_key.push(payload);
+            license_key.payload.push(payload);
+        }
+        license_key.properties.payload_size = license_key.payload.len();
+
+        // Create checksum
+        let checksum = adler32_checksum(
+            serialized_license_key.clone(),
+            self.checksum_init.0,
+            self.checksum_init.1,
+        );
+        for byte in checksum.to_be_bytes().iter() {
+            serialized_license_key.push(*byte);
+            license_key.checksum.push(*byte);
+        }
+        license_key.properties.checksum_size = checksum.to_be_bytes().len();
+
+        license_key.serialized_key = serialized_license_key.clone();
+        license_key
+    }
+
+    #[inline(always)]
+    pub fn validate_license_key(&self, key: &LicenseKey) -> LicenseKeyStatus {
+        let license_key = LicenseKey::deserialize(
+            &key.serialized_key,
+            key.properties.key_size,
+            key.properties.payload_size,
+            key.properties.checksum_size,
+        )
+        .unwrap_or_else(|_| LicenseKey::default());
+
+        // Validate checksum
+        let mut checksum_bytes = Vec::new();
+        checksum_bytes.extend(license_key.key.clone());
+        checksum_bytes.extend(license_key.payload.clone());
+
+        let checksum = adler32_checksum(checksum_bytes, self.checksum_init.0, self.checksum_init.1)
+            .to_be_bytes()
+            .to_vec();
+        if checksum != license_key.checksum {
+            return LicenseKeyStatus::Invalid;
+        }
+
+        // Validate seed from blacklist
+        for bl in self.blacklist.iter() {
```

Comment:
Maybe somthing like [try_fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold) would be more suited. However you already know the keys that are blacklisted, so maybe store them in [HashSet](https://doc.rust-lang.org/std/collections/struct.HashSet.html)? Also I would suggest you to use much more Results. Don't try to write smort libraries, that asume what developer inteded to do.
```rust
pub fn validate_license_key(&self, key: &LicenseKey) -> Result<(), Box<dyn Error>> {
    let lk = LicenseKey::deserialize(...)?
    
    let checksum = adler32_checksum([lk.key, lk.payload].concat(), self.checksum_init.0, self.checksum_init.1);

    if checksum != lk.checksum {
        LkError::InvalidChecksum(...)?
    }

    if self.blacklist.contains(lk.key) {
        LkError::Blacklisted(...)?
    }

    // this is actually strange size check, I would suggest you gake max and then check length
    let max_value = self.byte_check.iter().map(|&(x, y)| x).max();
    if max_value > lk.key.len() {
        LkError::Invalid(...)?
    }

    Ok(())
}
```

Reply:
Thanks for help, I rewrite that a little bit on my own.